File: Reference/tests/Brontide.Reference.Studio.Tests/ComponentPublisherEvidenceTests.cs
using System.Security.Cryptography;
using System.Collections;
using System.Text;
using System.Text.Json;

namespace Brontide.Reference.Studio.Tests;

public sealed partial class ComponentBindingIntegrationTests
{
    private sealed class SwitchingArtifactFiles(IReadOnlyList<ProviderArtifactAcquisitionFile> original)
        : IReadOnlyList<ProviderArtifactAcquisitionFile>
    {
        private int _enumerations;

        public int Count => original.Count;

        public ProviderArtifactAcquisitionFile this[int index] => original[index];

        public IEnumerator<ProviderArtifactAcquisitionFile> GetEnumerator()
        {
            _enumerations++;
            return (_enumerations == 1
                ? original
                : original.Select((file, index) => index == 0 ? file with { Sha256 = new string('C', 64) } : file)
                    .ToArray()).GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }

    private sealed record Cbi34Observation(
        string Code,
        bool Verified,
        string PayloadSha256,
        string? PublisherKeyId,
        string TrustCode,
        string AdmissionCode);

    private static ProviderArtifactAcquisitionRequest Cbi34CanonicalRequest(JsonElement manifest)
    {
        var files = manifest.GetProperty("files").EnumerateArray()
            .Select(file => new ProviderArtifactAcquisitionFile(
                file.GetProperty("path").GetString()!,
                file.GetProperty("sha256").GetString()!,
                file.GetProperty("length").GetInt64()))
            .ToArray();
        return new(
            ProviderArtifactSourceId.Create("fixture://brontide/publisher-evidence"),
            ProviderArtifactSetId.Create(manifest.GetProperty("identity").GetString()!),
            files,
            manifest.GetProperty("executablePath").GetString()!,
            manifest.GetProperty("arguments").EnumerateArray().Select(value => value.GetString()!).ToArray(),
            files.Sum(file => file.Length));
    }

    private static ProviderArtifactAcquisitionRequest Cbi34Mutate(
        ProviderArtifactAcquisitionRequest request,
        string mutation)
    {
        var files = request.Files.ToArray();
        var executable = request.ExecutablePath;
        var arguments = request.Arguments.ToArray();
        if (mutation == "path")
        {
            var index = Array.FindIndex(files, file => file.RelativePath == "data/config.json");
            files[index] = files[index] with { RelativePath = "data/settings.json" };
        }
        else if (mutation == "digest")
        {
            files[0] = files[0] with { Sha256 = new string('C', 64) };
        }
        else if (mutation == "length")
        {
            files[0] = files[0] with { Length = files[0].Length + 1 };
        }
        else if (mutation == "executable")
        {
            executable = files.Single(file => file.RelativePath != request.ExecutablePath).RelativePath;
        }
        else if (mutation == "arguments")
        {
            arguments = [.. arguments, "--changed"];
        }

        var identity = ProviderArtifactSetIdentity.Compute(
            files.Select(file => new ProviderArtifactFile(file.RelativePath, file.Sha256)), executable, arguments);
        return request with
        {
            Identity = identity,
            Files = files,
            ExecutablePath = executable,
            Arguments = arguments,
            MaxTotalBytes = files.Sum(file => file.Length),
        };
    }

    private static ProviderPublisherEvidence Cbi34Sign(
        ProviderArtifactAcquisitionRequest request,
        ECDsa key,
        string mutation)
    {
        var publicKey = key.ExportSubjectPublicKeyInfo();
        var keyId = ProviderPublisherKeyId.Create(Convert.ToHexString(SHA256.HashData(publicKey)));
        var signature = key.SignData(
            ProviderArtifactPublisherManifest.Encode(request),
            HashAlgorithmName.SHA256,
            DSASignatureFormat.Rfc3279DerSequence);
        var evidence = new ProviderPublisherEvidence(
            keyId,
            "ECDSA-P256-SHA256",
            Convert.ToBase64String(publicKey),
            Convert.ToBase64String(signature));
        return mutation switch
        {
            "absent" => evidence,
            "algorithm" => evidence with { Algorithm = "unknown" },
            "key" => evidence with { PublicKeySpkiBase64 = "not-base64" },
            "key-id" => evidence with { PublisherKeyId = ProviderPublisherKeyId.Create(new string('0', 64)) },
            "signature" => evidence with
            {
                SignatureBase64 = Convert.ToBase64String(key.SignData(
                    Encoding.UTF8.GetBytes("different payload"),
                    HashAlgorithmName.SHA256,
                    DSASignatureFormat.Rfc3279DerSequence)),
            },
            _ => evidence,
        };
    }

    private static Cbi34Observation Cbi34Run(JsonElement manifest, JsonElement vector)
    {
        var mutation = vector.GetProperty("mutation").GetString()!;
        var original = Cbi34CanonicalRequest(manifest);
        using var key = ECDsa.Create(ECCurve.NamedCurves.nistP256);
        var evidence = Cbi34Sign(original, key, mutation);
        var request = mutation is "path" or "digest" or "length" or "executable" or "arguments"
            ? Cbi34Mutate(original, mutation)
            : original;
        var result = ProviderArtifactPublisherEvidenceVerifier.Verify(
            request,
            mutation == "absent" ? null : evidence);
        return new(
            result.Code,
            result.IsVerified,
            result.PayloadSha256,
            result.PublisherKeyId?.Value,
            result.TrustCode,
            result.AdmissionCode);
    }

    private static async Task<JsonDocument> Cbi34FixtureAsync() => JsonDocument.Parse(
        await File.ReadAllTextAsync(Path.Combine(
            TestContext.CurrentContext.TestDirectory,
            "component-management",
            "fixtures",
            "cbi34-publisher-evidence-vectors.json")));

    [Test]
    public async Task Shared_cbi34_vectors_verify_publisher_evidence_without_granting_trust()
    {
        using var fixture = await Cbi34FixtureAsync();
        var manifest = fixture.RootElement.GetProperty("canonicalManifest");
        foreach (var vector in fixture.RootElement.GetProperty("vectors").EnumerateArray())
        {
            var observation = Cbi34Run(manifest, vector);
            Assert.Multiple(() =>
            {
                Assert.That(observation.Code, Is.EqualTo(vector.GetProperty("code").GetString()));
                Assert.That(observation.Verified, Is.EqualTo(vector.GetProperty("verified").GetBoolean()));
                Assert.That(observation.TrustCode, Is.EqualTo("publisher-trust-not-evaluated"));
                Assert.That(observation.AdmissionCode, Is.EqualTo("admission-not-attempted"));
            });
        }
    }

    [Test]
    public async Task Cbi34_C1_canonical_payload_covers_the_complete_acquisition_manifest()
    {
        using var fixture = await Cbi34FixtureAsync();
        var manifest = fixture.RootElement.GetProperty("canonicalManifest");
        var request = Cbi34CanonicalRequest(manifest);
        Assert.That(ProviderArtifactPublisherManifest.Digest(request),
            Is.EqualTo(manifest.GetProperty("payloadSha256").GetString()));
        Assert.That(ProviderArtifactPublisherManifest.Encode(request with
            {
                ExpectedSource = ProviderArtifactSourceId.Create("fixture://brontide/other-source"),
                MaxTotalBytes = request.MaxTotalBytes + 100,
            }), Is.EqualTo(ProviderArtifactPublisherManifest.Encode(request)));
        Assert.That(ProviderArtifactPublisherManifest.Digest(request with
            {
                Files = new SwitchingArtifactFiles(request.Files),
            }), Is.EqualTo(manifest.GetProperty("payloadSha256").GetString()));
    }

    [Test]
    public async Task Cbi34_C2_evidence_has_an_explicit_key_identity_and_algorithm()
    {
        using var fixture = await Cbi34FixtureAsync();
        var manifest = fixture.RootElement.GetProperty("canonicalManifest");
        foreach (var id in new[] { "cbi34-02-not-provided", "cbi34-03-unsupported", "cbi34-04-malformed-key", "cbi34-05-key-id-mismatch" })
        {
            var vector = fixture.RootElement.GetProperty("vectors").EnumerateArray()
                .Single(item => item.GetProperty("id").GetString() == id);
            var observation = Cbi34Run(manifest, vector);
            Assert.That(observation.Code, Is.EqualTo(vector.GetProperty("code").GetString()));
            Assert.That(observation.Verified, Is.False);
        }

        var request = Cbi34CanonicalRequest(manifest);
        using var wrongCurve = ECDsa.Create(ECCurve.NamedCurves.nistP384);
        var wrongCurveEvidence = Cbi34Sign(request, wrongCurve, "none");
        Assert.That(ProviderArtifactPublisherEvidenceVerifier.Verify(request, wrongCurveEvidence).Code,
            Is.EqualTo("publisher-evidence-malformed"));
    }

    [Test]
    public async Task Cbi34_C3_verification_binds_the_exact_canonical_payload()
    {
        using var fixture = await Cbi34FixtureAsync();
        var manifest = fixture.RootElement.GetProperty("canonicalManifest");
        foreach (var vector in fixture.RootElement.GetProperty("vectors").EnumerateArray()
                     .Where(item => item.GetProperty("id").GetString() is
                         "cbi34-06-signature-changed" or "cbi34-07-path-changed" or
                         "cbi34-08-digest-changed" or "cbi34-09-length-changed" or
                         "cbi34-10-executable-changed" or "cbi34-11-arguments-changed"))
        {
            Assert.That(Cbi34Run(manifest, vector).Code, Is.EqualTo("publisher-evidence-invalid"));
        }
    }

    [Test]
    public async Task Cbi34_C4_validity_is_not_trust_or_admission()
    {
        using var fixture = await Cbi34FixtureAsync();
        var valid = fixture.RootElement.GetProperty("vectors").EnumerateArray().First();
        var observation = Cbi34Run(fixture.RootElement.GetProperty("canonicalManifest"), valid);
        Assert.Multiple(() =>
        {
            Assert.That(observation.Verified, Is.True);
            Assert.That(observation.TrustCode, Is.EqualTo("publisher-trust-not-evaluated"));
            Assert.That(observation.AdmissionCode, Is.EqualTo("admission-not-attempted"));
        });
    }

    [Test]
    [Category("CrossProcess")]
    public void Cbi34_C5_explicit_caller_policy_may_compose_valid_evidence_with_cbi33()
    {
        var testRoot = Path.Combine(Path.GetTempPath(), $"brontide-cbi34-{Guid.NewGuid():N}");
        try
        {
            var (request, source) = Cbi33Input("reference", "none");
            using var key = ECDsa.Create(ECCurve.NamedCurves.nistP256);
            var verification = ProviderArtifactPublisherEvidenceVerifier.Verify(request, Cbi34Sign(request, key, "none"));
            Assert.That(verification.IsVerified, Is.True);
            var store = new ContentAddressedProviderStore(Path.Combine(testRoot, "store"));
            var acquisition = new ProviderArtifactAcquirer(store, Path.Combine(testRoot, "transactions"))
                .Acquire(request, source);
            Assert.Multiple(() =>
            {
                Assert.That(verification.AdmissionCode, Is.EqualTo("admission-not-attempted"));
                Assert.That(acquisition.TransportCode, Is.EqualTo("transport-completed"));
                Assert.That(acquisition.AdmissionCode, Is.EqualTo("staged"));
            });
            Assert.That(store.Remove(request.Identity).Code, Is.EqualTo("removed"));
        }
        finally
        {
            Cbi32DeleteTree(testRoot);
        }
    }

    [Test]
    public async Task Cbi34_C6_both_roots_agree_on_portable_observations()
    {
        using var fixture = await Cbi34FixtureAsync();
        var manifest = fixture.RootElement.GetProperty("canonicalManifest");
        var vectors = fixture.RootElement.GetProperty("vectors").EnumerateArray().ToArray();
        Assert.That(vectors.Select(vector => Cbi34Run(manifest, vector).Code),
            Is.EqualTo(vectors.Select(vector => vector.GetProperty("code").GetString())));
    }
}

File: Minimal/tests/Brontide.Minimal.Host.Tests/ComponentPublisherEvidenceTests.fs
namespace Brontide.Minimal.Host.Tests

open System
open System.IO
open System.Security.Cryptography
open System.Text
open System.Text.Json
open Brontide.Minimal.Host
open NUnit.Framework

type private Cbi34Observation =
    { Code: string
      Verified: bool
      PayloadSha256: string
      PublisherKeyId: string option
      TrustCode: string
      AdmissionCode: string }

type private Cbi34MemorySource(identity: ProviderArtifactSourceId, members: Map<string, byte array>) =
    interface IProviderArtifactSource with
        member _.Identity = identity
        member _.OpenRead path =
            members |> Map.tryFind path |> Option.map (fun bytes -> new MemoryStream(bytes, false) :> Stream)

[<TestFixture>]
type ComponentPublisherEvidenceTests() =
    let required (value: string | null) =
        match value with
        | null -> failwith "A CBI34 fixture value was missing."
        | present -> present

    let fixture () =
        JsonDocument.Parse(File.ReadAllText(Path.Combine(
            TestContext.CurrentContext.TestDirectory,
            "component-management",
            "fixtures",
            "cbi34-publisher-evidence-vectors.json")))

    let canonicalRequest (manifest: JsonElement) =
        let files =
            manifest.GetProperty("files").EnumerateArray()
            |> Seq.map (fun file ->
                { RelativePath = file.GetProperty("path").GetString() |> required
                  Sha256 = file.GetProperty("sha256").GetString() |> required
                  Length = file.GetProperty("length").GetInt64() }: ProviderArtifactAcquisitionFile)
            |> Seq.toList
        { ExpectedSource = ProviderArtifactSourceId.create "fixture://brontide/publisher-evidence"
          Identity = manifest.GetProperty("identity").GetString() |> required |> ProviderArtifactSetId.create
          Files = files
          ExecutablePath = manifest.GetProperty("executablePath").GetString() |> required
          Arguments = manifest.GetProperty("arguments").EnumerateArray() |> Seq.map (fun value -> value.GetString() |> required) |> Seq.toList
          MaxTotalBytes = files |> List.sumBy _.Length }: ProviderArtifactAcquisitionRequest

    let mutate (request: ProviderArtifactAcquisitionRequest) mutation =
        let files =
            match mutation with
            | "path" -> request.Files |> List.map (fun file ->
                if file.RelativePath = "data/config.json" then { file with RelativePath = "data/settings.json" } else file)
            | "digest" -> { request.Files.Head with Sha256 = String('C', 64) } :: request.Files.Tail
            | "length" -> { request.Files.Head with Length = request.Files.Head.Length + 1L } :: request.Files.Tail
            | _ -> request.Files
        let executable =
            if mutation = "executable" then files |> List.find (fun file -> file.RelativePath <> request.ExecutablePath) |> _.RelativePath
            else request.ExecutablePath
        let arguments = if mutation = "arguments" then request.Arguments @ [ "--changed" ] else request.Arguments
        let artifacts = files |> List.map (fun file ->
            { RelativePath = file.RelativePath; Sha256 = file.Sha256 }: ProviderArtifactFile)
        { request with
            Identity = ProviderArtifactSetIdentity.compute artifacts executable arguments
            Files = files
            ExecutablePath = executable
            Arguments = arguments
            MaxTotalBytes = files |> List.sumBy _.Length }

    let sign (request: ProviderArtifactAcquisitionRequest) (key: ECDsa) mutation =
        let publicKey = key.ExportSubjectPublicKeyInfo()
        let keyId = SHA256.HashData publicKey |> Convert.ToHexString |> ProviderPublisherKeyId.create
        let signature = key.SignData(
            ProviderArtifactPublisherManifest.encode request,
            HashAlgorithmName.SHA256,
            DSASignatureFormat.Rfc3279DerSequence)
        let evidence =
            { PublisherKeyId = keyId
              Algorithm = "ECDSA-P256-SHA256"
              PublicKeySpkiBase64 = Convert.ToBase64String publicKey
              SignatureBase64 = Convert.ToBase64String signature }
        match mutation with
        | "algorithm" -> { evidence with Algorithm = "unknown" }
        | "key" -> { evidence with PublicKeySpkiBase64 = "not-base64" }
        | "key-id" -> { evidence with PublisherKeyId = ProviderPublisherKeyId.create (String('0', 64)) }
        | "signature" ->
            let changed = key.SignData(
                Encoding.UTF8.GetBytes "different payload",
                HashAlgorithmName.SHA256,
                DSASignatureFormat.Rfc3279DerSequence)
            { evidence with SignatureBase64 = Convert.ToBase64String changed }
        | _ -> evidence

    let run (manifest: JsonElement) (vector: JsonElement) =
        let mutation = vector.GetProperty("mutation").GetString() |> required
        let original = canonicalRequest manifest
        use key = ECDsa.Create(ECCurve.NamedCurves.nistP256)
        let evidence = sign original key mutation
        let request =
            if [ "path"; "digest"; "length"; "executable"; "arguments" ] |> List.contains mutation then
                mutate original mutation
            else original
        let result = ProviderArtifactPublisherEvidenceVerifier.verify request (if mutation = "absent" then None else Some evidence)
        { Code = result.Code
          Verified = result.IsVerified
          PayloadSha256 = result.PayloadSha256
          PublisherKeyId = result.PublisherKeyId |> Option.map ProviderPublisherKeyId.value
          TrustCode = result.TrustCode
          AdmissionCode = result.AdmissionCode }

    [<Test>]
    member _.``shared CBI34 vectors verify publisher evidence without granting trust``() =
        use document = fixture ()
        let manifest = document.RootElement.GetProperty("canonicalManifest")
        for vector in document.RootElement.GetProperty("vectors").EnumerateArray() do
            let observation = run manifest vector
            Assert.Multiple(Action(fun () ->
                Assert.That(observation.Code, Is.EqualTo(vector.GetProperty("code").GetString()))
                Assert.That(observation.Verified, Is.EqualTo(vector.GetProperty("verified").GetBoolean()))
                Assert.That(observation.TrustCode, Is.EqualTo("publisher-trust-not-evaluated"))
                Assert.That(observation.AdmissionCode, Is.EqualTo("admission-not-attempted"))))

    [<Test>]
    member _.``CBI34 C1 canonical payload covers the complete acquisition manifest``() =
        use document = fixture ()
        let manifest = document.RootElement.GetProperty("canonicalManifest")
        Assert.That(
            ProviderArtifactPublisherManifest.digest (canonicalRequest manifest),
            Is.EqualTo(manifest.GetProperty("payloadSha256").GetString()))
        let request = canonicalRequest manifest
        let hostPolicyChanged =
            { request with
                ExpectedSource = ProviderArtifactSourceId.create "fixture://brontide/other-source"
                MaxTotalBytes = request.MaxTotalBytes + 100L }
        Assert.That(ProviderArtifactPublisherManifest.encode hostPolicyChanged,
                    Is.EqualTo(ProviderArtifactPublisherManifest.encode request :> obj))

    [<Test>]
    member _.``CBI34 C2 evidence has an explicit key identity and algorithm``() =
        use document = fixture ()
        let manifest = document.RootElement.GetProperty("canonicalManifest")
        let ids = Set [ "cbi34-02-not-provided"; "cbi34-03-unsupported"; "cbi34-04-malformed-key"; "cbi34-05-key-id-mismatch" ]
        for vector in document.RootElement.GetProperty("vectors").EnumerateArray() do
            if ids.Contains(vector.GetProperty("id").GetString() |> required) then
                let observation = run manifest vector
                Assert.That(observation.Code, Is.EqualTo(vector.GetProperty("code").GetString()))
                Assert.That(observation.Verified, Is.False)
        let request = canonicalRequest manifest
        use wrongCurve = ECDsa.Create(ECCurve.NamedCurves.nistP384)
        let wrongCurveEvidence = sign request wrongCurve "none"
        Assert.That(
            (ProviderArtifactPublisherEvidenceVerifier.verify request (Some wrongCurveEvidence)).Code,
            Is.EqualTo("publisher-evidence-malformed"))

    [<Test>]
    member _.``CBI34 C3 verification binds the exact canonical payload``() =
        use document = fixture ()
        let manifest = document.RootElement.GetProperty("canonicalManifest")
        for vector in document.RootElement.GetProperty("vectors").EnumerateArray() do
            let id = vector.GetProperty("id").GetString() |> required
            if id.CompareTo("cbi34-06") >= 0 then
                Assert.That((run manifest vector).Code, Is.EqualTo("publisher-evidence-invalid"))

    [<Test>]
    member _.``CBI34 C4 validity is not trust or admission``() =
        use document = fixture ()
        let observation = run (document.RootElement.GetProperty("canonicalManifest"))
                              (document.RootElement.GetProperty("vectors")[0])
        Assert.Multiple(Action(fun () ->
            Assert.That(observation.Verified, Is.True)
            Assert.That(observation.TrustCode, Is.EqualTo("publisher-trust-not-evaluated"))
            Assert.That(observation.AdmissionCode, Is.EqualTo("admission-not-attempted"))))

    [<Test>]
    member _.``CBI34 C5 explicit caller policy may compose valid evidence with CBI33``() =
        let bytes = Encoding.UTF8.GetBytes "provider"
        let digest = SHA256.HashData bytes |> Convert.ToHexString
        let files: ProviderArtifactAcquisitionFile list =
            [ { RelativePath = "provider.bin"; Sha256 = digest; Length = int64 bytes.Length } ]
        let artifacts: ProviderArtifactFile list =
            [ { RelativePath = "provider.bin"; Sha256 = digest } ]
        let sourceId = ProviderArtifactSourceId.create "fixture://brontide/composition"
        let request: ProviderArtifactAcquisitionRequest =
            { ExpectedSource = sourceId
              Identity = ProviderArtifactSetIdentity.compute artifacts "provider.bin" []
              Files = files
              ExecutablePath = "provider.bin"
              Arguments = []
              MaxTotalBytes = int64 bytes.Length }
        use key = ECDsa.Create(ECCurve.NamedCurves.nistP256)
        let verification = ProviderArtifactPublisherEvidenceVerifier.verify request (Some(sign request key "none"))
        let testRoot = Path.Combine(Path.GetTempPath(), $"brontide-cbi34-{Guid.NewGuid():N}")
        try
            let store = ContentAddressedProviderStore(Path.Combine(testRoot, "store"))
            let source = Cbi34MemorySource(sourceId, Map [ "provider.bin", bytes ])
            let acquisition = ProviderArtifactAcquirer(store, Path.Combine(testRoot, "transactions")).Acquire(request, source)
            Assert.Multiple(Action(fun () ->
                Assert.That(verification.IsVerified, Is.True)
                Assert.That(verification.AdmissionCode, Is.EqualTo("admission-not-attempted"))
                Assert.That(acquisition.TransportCode, Is.EqualTo("transport-completed"))
                Assert.That(acquisition.AdmissionCode, Is.EqualTo("staged"))))
            Assert.That(store.Remove(request.Identity).Code, Is.EqualTo("removed"))
        finally
            if Directory.Exists testRoot then
                Directory.EnumerateFiles(testRoot, "*", SearchOption.AllDirectories)
                |> Seq.iter (fun file -> File.SetAttributes(file, FileAttributes.Normal))
                Directory.Delete(testRoot, true)

    [<Test>]
    member _.``CBI34 C6 both roots agree on portable observations``() =
        use document = fixture ()
        let manifest = document.RootElement.GetProperty("canonicalManifest")
        let actual = document.RootElement.GetProperty("vectors").EnumerateArray() |> Seq.map (run manifest >> _.Code) |> Seq.toList
        let expected: string list =
            document.RootElement.GetProperty("vectors").EnumerateArray()
            |> Seq.map (fun vector -> vector.GetProperty("code").GetString() |> required)
            |> Seq.toList
        Assert.That(actual, Is.EqualTo(expected :> obj))
